Frameworks/CoreGraphics/CGFont.mm

```diff
@@ -16,615 +16,134 @@
 
 #import <StubReturn.h>
 #import <Starboard.h>
-#import <algorithm>
-#import <vector>
-#import <map>
-#import "CGContextInternal.h"
-#import "CGFontInternal.h"
-#import "UIFontInternal.h"
-#import "Etc.h"
-
-extern "C" {
-#import <ft2build.h>
-#import FT_FREETYPE_H
-#import <ftglyph.h>
-#import <tttables.h>
-#import <ftadvanc.h>
-#import <ftsizes.h>
-#import <ftsnames.h>
-#import <ttnameid.h>
-}
+
+#import <CFRuntime.h>
+#import <CoreGraphics/DWriteWrapper.h>
+
+#include <COMIncludes.h>
+#import <DWrite_3.h>
+#import <wrl/client.h>
+#include <COMIncludes_End.h>
+
+#import "LoggingNative.h"
 
 const CFStringRef kCGFontVariationAxisName = static_cast<CFStringRef>(@"kCGFontVariationAxisName");
 const CFStringRef kCGFontVariationAxisMinValue = static_cast<CFStringRef>(@"kCGFontVariationAxisMinValue");
 const CFStringRef kCGFontVariationAxisMaxValue = static_cast<CFStringRef>(@"kCGFontVariationAxisMaxValue");
 const CFStringRef kCGFontVariationAxisDefaultValue = static_cast<CFStringRef>(@"kCGFontVariationAxisDefaultValue");
 
-DWORD CGFontGetFontBBox(CGRect* ret, id font);
-
-static IWLazyClassLookup _LazyUIFont("UIFont");
+using namespace Microsoft::WRL;
 
-#include "LoggingNative.h"
+static const wchar_t* g_logTag = L"CGFont";
 
-static const wchar_t* TAG = L"CGFont";
+struct __CGFont {
+    CFRuntimeBase _base;
+    ComPtr<IDWriteFontFace> _dwriteFontFace;
 
-static float spacing = 1.0f;
+    struct DWRITE_FONT_METRICS _metrics;
+    bool _cachedMetrics; // Set to true when _metrics is init'd
+};
 
-static DWORD textureBuffer[1024];
-static DWORD destBuffer[1024];
+static Boolean __CGFontEqual(CFTypeRef cf1, CFTypeRef cf2) {
+    struct __CGFont* font1 = (struct __CGFont*)cf1;
+    struct __CGFont* font2 = (struct __CGFont*)cf2;
 
-#define USE_CAIRO
-
-void CGFontSetFTFontSize(id uiFont, void* ftFont, float pointSize, float scale) {
-    UIFont* fontInfo = uiFont;
-
-    FT_Set_Char_Size((FT_Face)ftFont, 0, (FT_F26Dot6)(pointSize * 64.0f), (FT_UInt)(72.f * scale), (FT_UInt)(72.f * scale));
+    return (font1->_dwriteFontFace == font2->_dwriteFontFace);
```

Comment:
they're an optimization to avoid calling getters on _dwriteFontFace every time, they don't matter in terms of comparison.

---
In reply to: [83337659](https://github.com/Microsoft/WinObjC/pull/1145#discussion_r83337659) [](ancestors = 83337659)